File: src/test/java/tech/sirwellington/alchemy/http/Step4ImplTest.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http;

import java.net.URL;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnSuccess;
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.Step4;
import tech.sirwellington.alchemy.test.junit.runners.AlchemyTestRunner;
import tech.sirwellington.alchemy.test.junit.runners.Repeat;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows;

/**
 * @author SirWellington
 */
@RunWith(AlchemyTestRunner.class)
public class Step4ImplTest
{
    @Mock
    private AlchemyHttpStateMachine stateMachine;

    @Mock
    private HttpRequest request;

    @Captor
    private ArgumentCaptor<HttpRequest> requestCaptor;

    @Mock
    private OnSuccess<TestPojo> onSuccess;

    private Class<TestPojo> responseClass;

    private Step4<TestPojo> instance;

    @Before
    public void setUp()
    {
        responseClass = TestPojo.class;

        instance = new Step4Impl<>(stateMachine, request, responseClass);
        verifyNoInteractions(stateMachine);
    }

    @Repeat
    @Test
    public void testAt()
    {
        URL url = one(Generators.validUrls());

        instance.at(url);

        verify(stateMachine).executeSync(requestCaptor.capture(), eq(responseClass));

        var requestMade = requestCaptor.getValue();
        assertThat(requestMade, notNullValue());
        assertThat(requestMade, not(sameInstance(request)));
        assertThat(requestMade.url(), equalTo(url));
    }

    @Test
    public void testAtWithBadArgs() throws Exception
    {
        assertThrows(() -> instance.at(""))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testOnSuccess()
    {
        // Edge cases
        instance.onSuccess(onSuccess);

        verify(stateMachine).jumpToStep5(request, responseClass, onSuccess);
    }

    @Test
    public void testToString()
    {
        assertThat(Strings.isNullOrEmpty(instance.toString()), equalTo(false));
    }
}

File: src/test/java/tech/sirwellington/alchemy/http/Step4ImplTest.kt
/*
 * Copyright © 2018. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http

import com.nhaarman.mockito_kotlin.KArgumentCaptor
import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.eq
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.not
import org.hamcrest.Matchers.notNullValue
import org.hamcrest.Matchers.sameInstance
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyZeroInteractions
import tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnSuccess
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.Step4
import tech.sirwellington.alchemy.http.Generators.validUrls
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.AlchemyTestRunner
import tech.sirwellington.alchemy.test.junit.runners.Repeat

/**
 *
 * @author SirWellington
 */
@RunWith(AlchemyTestRunner::class)
class Step4ImplTest
{

    @Mock
    private lateinit var stateMachine: AlchemyHttpStateMachine

    @Mock
    private lateinit var request: HttpRequest

    private lateinit var requestCaptor: KArgumentCaptor<HttpRequest>

    @Mock
    private lateinit var onSuccess: OnSuccess<TestPojo>

    private lateinit var responseClass: Class<TestPojo>

    private lateinit var instance: Step4<TestPojo>

    @Before
    fun setUp()
    {
        responseClass = TestPojo::class.java

        instance = Step4Impl(stateMachine, request, responseClass)
        verifyZeroInteractions(stateMachine)

        requestCaptor = argumentCaptor()
    }

    @Repeat
    @Test
    fun testAt()
    {
        val url = one(validUrls())

        instance.at(url)

        verify(stateMachine).executeSync(requestCaptor.capture(), eq(responseClass))

        val requestMade = requestCaptor.firstValue
        assertThat(requestMade, notNullValue())
        assertThat(requestMade, not(sameInstance(request)))
        assertThat(requestMade.url, equalTo(url))
    }

    @Test
    @Throws(Exception::class)
    fun testAtWithBadArgs()
    {
        assertThrows { instance.at("") }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testOnSuccess()
    {
        //Edge cases
        instance.onSuccess(onSuccess)

        verify(stateMachine).jumpToStep5(request, responseClass, onSuccess)
    }

    @Test
    fun testToString()
    {
        assertThat(Strings.isNullOrEmpty(instance.toString()), equalTo(false))
    }

}
